### PR TITLE
[PySpark] add gpu support for spark local mode

### PR DESCRIPTION
**Pull request by wbo4958**
This PR is to support gpu training (use_gpu=True) on Spark local mode.

Please note that, 

1. Spark local mode does not support GPU configuration, so it will fail if the user adds related GPU configurations.
2. For now, we only support num_workers=1 when running on GPU on Spark local mode.

This feature is needed since it's really convenient for local debugging.

### CODE REVIEW

**Comment by wbo4958** [comment]
@trivialfis How to run python tests locally? is there a doc for that?

**Comment by trivialfis** [comment]
```
pytest ./tests/python
```
You can check the doc for pytest.

File: python-package/xgboost/spark/core.py [hunk 1]
```diff
@@ -281,19 +283,40 @@ def _validate_params(self):
...
+                        "The spark cluster does not support gpu configuration for local mode. "
+                        + "Please delete spark.executor.resource.gpu.amount and "
+                        + "spark.task.resource.gpu.amount"
+                    )
+
+                if self.getOrDefault(self.num_workers) > 1:
+                    raise ValueError(
+                        "Training XGBoost on the spark local mode only supports num_workers = 1, "
+                        + "and only primary GPU device will be used."
+                    )
```
Comment (WeichenXu123) [inline python-package/xgboost/spark/core.py:304]: I think we can still support multiple GPU workers in spark local mode.
We can get partition id from spark TaskContext, and use the partition id as the gpu_id for the corresponding spark task. WDYT ?

Comment (wbo4958) [reply python-package/xgboost/spark/core.py:304]: Thx. Hmm, there will be only 1 process for spark local mode, If we support the scenario of num_workers > 1, then all the GPU training tasks will run on the same process, which seems not to be supported due to the NCCL issue. @trivialfis Could you confirm that?

Comment (trivialfis) [reply python-package/xgboost/spark/core.py:304]: It should work but let's not invite trouble.

Comment (wbo4958) [reply python-package/xgboost/spark/core.py:304]: Yeah, I agree. The GPU supporting for the local mode is mostly used in local debugging.

**Review by WeichenXu123 — APPROVED** [review]

File: python-package/xgboost/spark/core.py (hunk 1 above)
Comment (WeichenXu123) [reply python-package/xgboost/spark/core.py:304]: @wbo4958 
A correction:
> local mode pyspark all the GPU training tasks will run on the same process

This is not true.
For pyspark, in barrier mode pyspark task, each task will be run in an individual python process.

Comment (wbo4958) [reply python-package/xgboost/spark/core.py:304]: Oh right, my bad. I previously thought it was the JVM side. Ok, let me add it support.

**Review by trivialfis — APPROVED** [review]

**Comment by trivialfis** [comment]
Please address the linter error.

**Pull request merged**